compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -45,19 +45,21 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def Apply(fn: Tree, args: List[Tree])(using Context): Apply = fn match
     case Block(Nil, expr) =>
       Apply(expr, args)
-    case _: RefTree | _: GenericApply | _: Inlined | _: Hole =>
-      ta.assignType(untpd.Apply(fn, args), fn, args)
     case _ =>
-      assert(ctx.reporter.errorsReported)
+      assert(
+        fn.isInstanceOf[RefTree | GenericApply | Inlined | Hole] || ctx.reporter.errorsReported,
+        s"Illegal Apply function prefix: $fn"
+      )
       ta.assignType(untpd.Apply(fn, args), fn, args)
 
   def TypeApply(fn: Tree, args: List[Tree])(using Context): TypeApply = fn match
     case Block(Nil, expr) =>
       TypeApply(expr, args)
-    case _: RefTree | _: GenericApply =>
-      ta.assignType(untpd.TypeApply(fn, args), fn, args)
     case _ =>
-      assert(ctx.reporter.errorsReported)
+      assert(
+        fn.isInstanceOf[RefTree | GenericApply] || ctx.reporter.errorsReported,
+        s"Illegal TypeApply function prefix: $fn"
```

Comment:
We should use the `i` interpolator, so it's pretty-printed.

Reply:
No. We need to see the exact AST. Otherwise we would not see the difference between a `TypeApply` and `Inlined`; or some similar cases where `show` is equivalent.

Reply:
Another way to see this is as a match error. Maybe I could encode it in a simpler way and make it fail with an equivalently useful output. I will try that.

Reply:
The alternative is
```scala
  def TypeApply(fn: Tree, args: List[Tree])(using Context): TypeApply = fn match
    case Block(Nil, expr) =>
      TypeApply(expr, args)
    case _: RefTree | _: GenericApply =>
      ta.assignType(untpd.TypeApply(fn, args), fn, args)
    case _ if ctx.reporter.errorsReported => // We expect the error. Otherwise, we want to throw a match error.
      ta.assignType(untpd.TypeApply(fn, args), fn, args)
```

But this one does duplicate `ta.assignType(untpd.TypeApply(fn, args), fn, args)`. I find the current one clearer.

Reply:
How about `i"Illegal TypeApply function prefix: $fn (${fn.className})"` then?  Don't need the verbose nested trees, I would think.

Reply:
Given the history of the failure, I would argue otherwise. Knowing which tree caused the issue in one of those large projects can be helpful in pinpointing the likely source of the problem.